README.md

```diff
@@ -1,22 +1,381 @@
 
-Code of interpreter for functional programming language based on the [writing an interpreter in Go](https://interpreterbook.com/) book.
-
-## TODO
-
-- [ ] floats
-- [x] counting rows
-- [ ] counting columns
-- [x] running with file as an argument
-- [x] one-line comments
-- [x] multiple-lines comments
-- [x] return statement only permitted in function's body
-- [x] return mandatory in function's body
-- [x] return statement can be empty if we don't want to return anything
-- [x] forbid reassigning constant
-- [x] forbid redeclaring constant in one block
-- [x] merge TRUE and FALSE tokens into one BOOLEAN token with different value
+Interpreter for programming language we named **Junior**.
+
+## What is it
+
+It's an interpreter written in Golang for programming language of our creation.<br />
+It's a project for *Theory of Compilation* classes at the *AGH* university.<br />
+It's based on [writing an interpreter in go](https://interpreterbook.com/) book.
+
+You can test it in action [here](https://junior-interpreter-online.herokuapp.com/).
+
+## How it Works
+
+It divides interpreting the Junior's code into 3 parts.
+
+1. Lexer - performing lexical analysis
+2. Parser - syntax analysis and building Abstract Syntax Tree.
+3. Evaluator - traversing AST and evaluating the program.
+
+|  | input | output |
+| :---: | :---: | :---: |
+| **Lexer** | Junior's program code | tokens |
+| **Parser** | tokens | Abstract Syntax Tree nodes |
+| **Evaluator** | Abstract Syntax Tree nodes | evaluated program statements |
+
+## Junior Language Specification
+
+Junior is an imperative programming language. It derive from functional programming paradigm.
+It's loosely typed but uses immutability. It has features like closures and IIFEs.
+It's based on [Monkey programming language](https://interpreterbook.com/#the-monkey-programming-language).
+
+Example program:
+
+```javascript
+const makeConcat = fun(y) {
+    return fun(x) {
+        return x + y;
+    };
+};
+
+const addWorld = makeConcat(" World!");
+
+addWorld("Hello"); // Hello World!
+```
+
+You can find more examples in the `examples` directory.
+
+### Grammar
+
+#### Keywords
+
+Reserved keywords of Junior:
+
+`const, fun, return, if, else, true, false`
+
+Reserved names of built-in functions:
+
+`print, last, first, rest, len, push`
+
+#### Statements
+
+Statements are instructions that Junior's program is contained of.
+In Junior statements are separated with semicolons.
+
+##### Const statement
+
+`const` `identifier` `=` `expression` `;`
+
+Const statement binds value evaluated from `expression` to variable `identifier` points to.
+Junior uses block scoping, there are three different kinds of scopes.
+1. global scope
+2. function scope
+3. if/else scope
+
+If variable is not found in the current scope the ancestor's scope is looked up, if interpreter fails to find given identifier even in the global scope a semantic error is evaluated.
```

Comment:
If a variable is not found in the current scope, the ancestor's scope is examined (?) taken into consideration ? xd
not sure what to put instead of look up cause it most often means searching in a book